File: classified/packages/agentserver/src/services/progressionTracker.ts
import { IAgentRuntime, logger, EventType } from '@elizaos/core';
import { CapabilityProgressionService } from './capabilityProgressionService';

export class ProgressionTracker {
  private runtime: IAgentRuntime;
  private progressionService: CapabilityProgressionService;

  constructor(runtime: IAgentRuntime, progressionService: CapabilityProgressionService) {
    this.runtime = runtime;
    this.progressionService = progressionService;
    this.setupEventListeners();
  }

  public stop(): void {
    // Event listeners are automatically cleaned up with the runtime
    // No manual cleanup needed for event-based approach
  }

  private setupEventListeners(): void {
    logger.info('[PROGRESSION_TRACKER] Setting up event-based progression tracking');

    // Skip setup in unlocked mode
    if (this.progressionService.isUnlockedModeEnabled()) {
      logger.info('[PROGRESSION_TRACKER] Skipping event listeners - unlocked mode enabled');
      return;
    }

    // Register event listeners for progression tracking
    this.runtime.registerEvent(EventType.SHELL_COMMAND_EXECUTED, async (params) => {
      await this.handleShellCommandExecuted(params);
    });

    this.runtime.registerEvent(EventType.AGENT_NAMED, async (params) => {
      await this.handleAgentNamed(params);
    });

    this.runtime.registerEvent(EventType.GOAL_CREATED, async (params) => {
      await this.handleGoalCreated(params);
    });

    this.runtime.registerEvent(EventType.TODO_CREATED, async (params) => {
      await this.handleTodoCreated(params);
    });

    this.runtime.registerEvent(EventType.BROWSER_ACTION_PERFORMED, async (params) => {
      await this.handleBrowserActionPerformed(params);
    });

    this.runtime.registerEvent(EventType.FORM_SUBMITTED, async (params) => {
      await this.handleFormSubmitted(params);
    });

    this.runtime.registerEvent(EventType.VISION_ACTION_PERFORMED, async (params) => {
      await this.handleVisionActionPerformed(params);
    });

    this.runtime.registerEvent(EventType.MICROPHONE_USED, async (params) => {
      await this.handleMicrophoneUsed(params);
    });

    this.runtime.registerEvent(EventType.CAPABILITY_USED, async (params) => {
      await this.handleCapabilityUsed(params);
    });

    logger.info('[PROGRESSION_TRACKER] Event listeners registered');
  }

  // Event handlers for progression tracking
  private async handleShellCommandExecuted(params: {
    command: string;
    exitCode: number;
    output?: string;
  }): Promise<void> {
    if (this.progressionService.isUnlockedModeEnabled()) {
      return;
    }

    if (params.exitCode === 0) {
      logger.info(
        '[PROGRESSION_TRACKER] Shell command executed successfully, recording capability usage'
      );
      await this.progressionService.recordCapabilityUsed('shell');
    }
  }

  private async handleAgentNamed(_params: { name: string }): Promise<void> {
    if (this.progressionService.isUnlockedModeEnabled()) {
      return;
    }

    logger.info(`[PROGRESSION_TRACKER] Agent named: ${_params.name}`);
    this.progressionService.recordAgentNamed(_params.name);
  }

  private async handleGoalCreated(_params: { goalData: Record<string, unknown> }): Promise<void> {
    if (this.progressionService.isUnlockedModeEnabled()) {
      return;
    }

    logger.info('[PROGRESSION_TRACKER] Goal created, recording capability usage');
    await this.progressionService.recordCapabilityUsed('goals');
  }

  private async handleTodoCreated(_params: { todoData: Record<string, unknown> }): Promise<void> {
    if (this.progressionService.isUnlockedModeEnabled()) {
      return;
    }

    logger.info('[PROGRESSION_TRACKER] Todo created, recording capability usage');
    await this.progressionService.recordCapabilityUsed('todo');
  }

  private async handleBrowserActionPerformed(params: {
    action: string;
    details?: Record<string, unknown>;
  }): Promise<void> {
    if (this.progressionService.isUnlockedModeEnabled()) {
      return;
    }

    logger.info(`[PROGRESSION_TRACKER] Browser action performed: ${params.action}`);
    await this.progressionService.recordCapabilityUsed('browser');
  }

  private async handleFormSubmitted(params: { details?: Record<string, unknown> }): Promise<void> {
    if (this.progressionService.isUnlockedModeEnabled()) {
      return;
    }

    logger.info('[PROGRESSION_TRACKER] Form submitted');
    await this.progressionService.recordFormSubmitted(params.details);
  }

  private async handleVisionActionPerformed(params: {
    action: string;
    details?: Record<string, unknown>;
  }): Promise<void> {
    if (this.progressionService.isUnlockedModeEnabled()) {
      return;
    }

    logger.info(`[PROGRESSION_TRACKER] Vision action performed: ${params.action}`);
    await this.progressionService.recordCapabilityUsed('vision');
  }

  private async handleMicrophoneUsed(_params: {
    details?: Record<string, unknown>;
  }): Promise<void> {
    if (this.progressionService.isUnlockedModeEnabled()) {
      return;
    }

    logger.info('[PROGRESSION_TRACKER] Microphone used');
    await this.progressionService.recordCapabilityUsed('microphone');
  }

  private async handleCapabilityUsed(params: {
    capability: string;
    details?: Record<string, unknown>;
  }): Promise<void> {
    if (this.progressionService.isUnlockedModeEnabled()) {
      return;
    }

    logger.info(`[PROGRESSION_TRACKER] Capability used: ${params.capability}`);
    await this.progressionService.recordCapabilityUsed(params.capability);
  }

  // Method to track shell command execution - called from API endpoints
  public async trackShellCommand(command: string, exitCode: number): Promise<void> {
    // Emit event instead of direct tracking
    await this.runtime.emitEvent(EventType.SHELL_COMMAND_EXECUTED, {
      command,
      exitCode,
    });
  }

  // Method to track goal creation - called from API endpoints
  public async trackGoalCreation(goalData: Record<string, unknown>): Promise<void> {
    // Emit event instead of direct tracking
    await this.runtime.emitEvent(EventType.GOAL_CREATED, {
      goalData,
    });
  }

  // Method to track todo creation - called from API endpoints
  public async trackTodoCreation(todoData: Record<string, unknown>): Promise<void> {
    // Emit event instead of direct tracking
    await this.runtime.emitEvent(EventType.TODO_CREATED, {
      todoData,
    });
  }

  // Method to track agent naming - should be called when agent name changes
  public async trackAgentNaming(name: string): Promise<void> {
    // Emit event instead of direct tracking
    await this.runtime.emitEvent(EventType.AGENT_NAMED, {
      name,
    });
  }

  // Method to manually track specific actions
  public async trackAction(actionType: string, details?: Record<string, unknown>): Promise<void> {
    logger.info(`[PROGRESSION_TRACKER] Manual action tracked: ${actionType}`, details);

    // Emit appropriate events based on action type
    switch (actionType) {
      case 'form_submitted':
        await this.runtime.emitEvent(EventType.FORM_SUBMITTED, { details });
        break;
      case 'browser_used':
        await this.runtime.emitEvent(EventType.BROWSER_ACTION_PERFORMED, {
          action: actionType,
          details,
        });
        break;
      case 'vision_used':
        await this.runtime.emitEvent(EventType.VISION_ACTION_PERFORMED, {
          action: actionType,
          details,
        });
        break;
      case 'microphone_used':
        await this.runtime.emitEvent(EventType.MICROPHONE_USED, { details });
        break;
      default:
        // For generic capability usage
        await this.runtime.emitEvent(EventType.CAPABILITY_USED, {
          capability: actionType,
          details,
        });
        break;
    }
  }

  // Method to check and display current progression status
  public getProgressionStatus(): Record<string, unknown> {
    const state = this.progressionService.getProgressionState();
    const unlockedCapabilities = this.progressionService.getUnlockedCapabilities();
    const availableLevels = this.progressionService.getAvailableLevels();
    const isUnlockedMode = this.progressionService.isUnlockedModeEnabled();

    return {
      mode: isUnlockedMode ? 'unlocked' : 'progression',
      isUnlockedMode,
      currentLevel: state.currentLevel,
      unlockedLevels: state.unlockedLevels,
      completedActions: state.completedActions,
      agentNamed: state.agentNamed,
      unlockedCapabilities,
      availableLevels: availableLevels.map((level) => ({
        id: level.id,
        name: level.name,
        description: level.description,
        isUnlocked: level.isUnlocked,
        capabilities: level.unlockedCapabilities,
      })),
    };
  }

  // Cleanup method to stop intervals
  public cleanup(): void {
    if (this.trackingInterval) {
      clearInterval(this.trackingInterval);
      this.trackingInterval = null;
    }
    if (this.namingInterval) {
      clearInterval(this.namingInterval);
      this.namingInterval = null;
    }
    logger.info('[PROGRESSION_TRACKER] Cleaned up tracking intervals');
  }
}

File: cpp_merge/classified/packages/agentserver/src/services/progressionTracker.cpp
#include "progressionTracker.hpp"

ProgressionTracker::ProgressionTracker(std::shared_ptr<IAgentRuntime> runtime, std::shared_ptr<CapabilityProgressionService> progressionService) {
    this->runtime = runtime;
    this->progressionService = progressionService;
    this->setupEventListeners();
}

void ProgressionTracker::stop()
{
}

void ProgressionTracker::setupEventListeners()
{
    logger->info(std:("[PROGRESSION_TRACKER] Setting up event-based progression tracking"));
    if (this->progressionService->isUnlockedModeEnabled()) {
        logger->info(std:("[PROGRESSION_TRACKER] Skipping event listeners - unlocked mode enabled"));
        return;
    }
    this->runtime->registerEvent(EventType->SHELL_COMMAND_EXECUTED, [=](auto params) mutable
    {
        std::async([=]() { this->handleShellCommandExecuted(params); });
    }
    );
    this->runtime->registerEvent(EventType->AGENT_NAMED, [=](auto params) mutable
    {
        std::async([=]() { this->handleAgentNamed(params); });
    }
    );
    this->runtime->registerEvent(EventType->GOAL_CREATED, [=](auto params) mutable
    {
        std::async([=]() { this->handleGoalCreated(params); });
    }
    );
    this->runtime->registerEvent(EventType->TODO_CREATED, [=](auto params) mutable
    {
        std::async([=]() { this->handleTodoCreated(params); });
    }
    );
    this->runtime->registerEvent(EventType->BROWSER_ACTION_PERFORMED, [=](auto params) mutable
    {
        std::async([=]() { this->handleBrowserActionPerformed(params); });
    }
    );
    this->runtime->registerEvent(EventType->FORM_SUBMITTED, [=](auto params) mutable
    {
        std::async([=]() { this->handleFormSubmitted(params); });
    }
    );
    this->runtime->registerEvent(EventType->VISION_ACTION_PERFORMED, [=](auto params) mutable
    {
        std::async([=]() { this->handleVisionActionPerformed(params); });
    }
    );
    this->runtime->registerEvent(EventType->MICROPHONE_USED, [=](auto params) mutable
    {
        std::async([=]() { this->handleMicrophoneUsed(params); });
    }
    );
    this->runtime->registerEvent(EventType->CAPABILITY_USED, [=](auto params) mutable
    {
        std::async([=]() { this->handleCapabilityUsed(params); });
    }
    );
    logger->info(std:("[PROGRESSION_TRACKER] Event listeners registered"));
}

std::shared_ptr<Promise<void>> ProgressionTracker::handleShellCommandExecuted(object params)
{
    if (this->progressionService->isUnlockedModeEnabled()) {
        return std::shared_ptr<Promise<void>>();
    }
    if (params["exitCode"] == 0) {
        logger->info(std:("[PROGRESSION_TRACKER] Shell command executed successfully, recording capability usage"));
        std::async([=]() { this->progressionService->recordCapabilityUsed(std:("shell")); });
    }
}

std::shared_ptr<Promise<void>> ProgressionTracker::handleAgentNamed(object _params)
{
    if (this->progressionService->isUnlockedModeEnabled()) {
        return std::shared_ptr<Promise<void>>();
    }
    logger->info(std:("[PROGRESSION_TRACKER] Agent named: ") + _params["name"] + string_empty);
    this->progressionService->recordAgentNamed(_params["name"]);
}

std::shared_ptr<Promise<void>> ProgressionTracker::handleGoalCreated(object _params)
{
    if (this->progressionService->isUnlockedModeEnabled()) {
        return std::shared_ptr<Promise<void>>();
    }
    logger->info(std:("[PROGRESSION_TRACKER] Goal created, recording capability usage"));
    std::async([=]() { this->progressionService->recordCapabilityUsed(std:("goals")); });
}

std::shared_ptr<Promise<void>> ProgressionTracker::handleTodoCreated(object _params)
{
    if (this->progressionService->isUnlockedModeEnabled()) {
        return std::shared_ptr<Promise<void>>();
    }
    logger->info(std:("[PROGRESSION_TRACKER] Todo created, recording capability usage"));
    std::async([=]() { this->progressionService->recordCapabilityUsed(std:("todo")); });
}

std::shared_ptr<Promise<void>> ProgressionTracker::handleBrowserActionPerformed(object params)
{
    if (this->progressionService->isUnlockedModeEnabled()) {
        return std::shared_ptr<Promise<void>>();
    }
    logger->info(std:("[PROGRESSION_TRACKER] Browser action performed: ") + params["action"] + string_empty);
    std::async([=]() { this->progressionService->recordCapabilityUsed(std:("browser")); });
}

std::shared_ptr<Promise<void>> ProgressionTracker::handleFormSubmitted(object params)
{
    if (this->progressionService->isUnlockedModeEnabled()) {
        return std::shared_ptr<Promise<void>>();
    }
    logger->info(std:("[PROGRESSION_TRACKER] Form submitted"));
    std::async([=]() { this->progressionService->recordFormSubmitted(params["details"]); });
}

std::shared_ptr<Promise<void>> ProgressionTracker::handleVisionActionPerformed(object params)
{
    if (this->progressionService->isUnlockedModeEnabled()) {
        return std::shared_ptr<Promise<void>>();
    }
    logger->info(std:("[PROGRESSION_TRACKER] Vision action performed: ") + params["action"] + string_empty);
    std::async([=]() { this->progressionService->recordCapabilityUsed(std:("vision")); });
}

std::shared_ptr<Promise<void>> ProgressionTracker::handleMicrophoneUsed(object _params)
{
    if (this->progressionService->isUnlockedModeEnabled()) {
        return std::shared_ptr<Promise<void>>();
    }
    logger->info(std:("[PROGRESSION_TRACKER] Microphone used"));
    std::async([=]() { this->progressionService->recordCapabilityUsed(std:("microphone")); });
}

std::shared_ptr<Promise<void>> ProgressionTracker::handleCapabilityUsed(object params)
{
    if (this->progressionService->isUnlockedModeEnabled()) {
        return std::shared_ptr<Promise<void>>();
    }
    logger->info(std:("[PROGRESSION_TRACKER] Capability used: ") + params["capability"] + string_empty);
    std::async([=]() { this->progressionService->recordCapabilityUsed(params["capability"]); });
}

std::shared_ptr<Promise<void>> ProgressionTracker::trackShellCommand(string command, double exitCode)
{
    std::async([=]() { this->runtime->emitEvent(EventType->SHELL_COMMAND_EXECUTED, object{
        object::pair{std:("command"), std:("command")}, 
        object::pair{std:("exitCode"), std:("exitCode")}
    }); });
    return std::shared_ptr<Promise<void>>();
}

std::shared_ptr<Promise<void>> ProgressionTracker::trackGoalCreation(Record<string, any> goalData)
{
    std::async([=]() { this->runtime->emitEvent(EventType->GOAL_CREATED, object{
        object::pair{std:("goalData"), std:("goalData")}
    }); });
    return std::shared_ptr<Promise<void>>();
}

std::shared_ptr<Promise<void>> ProgressionTracker::trackTodoCreation(Record<string, any> todoData)
{
    std::async([=]() { this->runtime->emitEvent(EventType->TODO_CREATED, object{
        object::pair{std:("todoData"), std:("todoData")}
    }); });
    return std::shared_ptr<Promise<void>>();
}

std::shared_ptr<Promise<void>> ProgressionTracker::trackAgentNaming(string name)
{
    std::async([=]() { this->runtime->emitEvent(EventType->AGENT_NAMED, object{
        object::pair{std:("name"), std:("name")}
    }); });
    return std::shared_ptr<Promise<void>>();
}

std::shared_ptr<Promise<void>> ProgressionTracker::trackAction(string actionType, Record<string, any> details)
{
    logger->info(std:("[PROGRESSION_TRACKER] Manual action tracked: ") + actionType + string_empty, details);
    static switch_type __switch6914_7798 = {
        { any(std:("form_submitted")), 1 },
        { any(std:("browser_used")), 2 },
        { any(std:("vision_used")), 3 },
        { any(std:("microphone_used")), 4 }
    };
    switch (__switch6914_7798[actionType])
    {
    case 1:
        std::async([=]() { this->runtime->emitEvent(EventType->FORM_SUBMITTED, object{
            object::pair{std:("details"), std:("details")}
        }); });
        break;
    case 2:
        std::async([=]() { this->runtime->emitEvent(EventType->BROWSER_ACTION_PERFORMED, object{
            object::pair{std:("action"), actionType}, 
            object::pair{std:("details"), std:("details")}
        }); });
        break;
    case 3:
        std::async([=]() { this->runtime->emitEvent(EventType->VISION_ACTION_PERFORMED, object{
            object::pair{std:("action"), actionType}, 
            object::pair{std:("details"), std:("details")}
        }); });
        break;
    case 4:
        std::async([=]() { this->runtime->emitEvent(EventType->MICROPHONE_USED, object{
            object::pair{std:("details"), std:("details")}
        }); });
        break;
    default:
        std::async([=]() { this->runtime->emitEvent(EventType->CAPABILITY_USED, object{
            object::pair{std:("capability"), actionType}, 
            object::pair{std:("details"), std:("details")}
        }); });
        break;
    }
    return std::shared_ptr<Promise<void>>();
}

Record<string, any> ProgressionTracker::getProgressionStatus()
{
    auto state = this->progressionService->getProgressionState();
    shared unlockedCapabilities = this->progressionService->getUnlockedCapabilities();
    auto availableLevels = this->progressionService->getAvailableLevels();
    auto isUnlockedMode = this->progressionService->isUnlockedModeEnabled();
    return object{
        object::pair{std:("mode"), (isUnlockedMode) ? std:("unlocked") : std:("progression")}, 
        object::pair{std:("isUnlockedMode"), std:("isUnlockedMode")}, 
        object::pair{std:("currentLevel"), state->currentLevel}, 
        object::pair{std:("unlockedLevels"), state->unlockedLevels}, 
        object::pair{std:("completedActions"), state->completedActions}, 
        object::pair{std:("agentNamed"), state->agentNamed}, 
        object::pair{std:("unlockedCapabilities"), std:("unlockedCapabilities")}, 
        object::pair{std:("availableLevels"), availableLevels->map([=](auto level) mutable
        {
            return (object{
                object::pair{std:("id"), level->id}, 
                object::pair{std:("name"), level->name}, 
                object::pair{std:("description"), level->description}, 
                object::pair{std:("isUnlocked"), level->isUnlocked}, 
                object::pair{std:("capabilities"), level->unlockedCapabilities}
            });
        }
        )}
    };
}

void ProgressionTracker::cleanup()
{
    if (this->trackingInterval) {
        clearInterval(this->trackingInterval);
        this->trackingInterval = nullptr;
    }
    if (this->namingInterval) {
        clearInterval(this->namingInterval);
        this->namingInterval = nullptr;
    }
    logger->info(std:("[PROGRESSION_TRACKER] Cleaned up tracking intervals"));
}

